listenbrainz_spark/stats/user.py

```diff
@@ -0,0 +1,214 @@
+                 , release_name
+                 , release_msid
+                 , release_mbid
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+                 , count(release_msid) as cnt
+              FROM %s
+          GROUP BY user_name, release_name, release_msid, release_mbid, artist_name, artist_msid, artist_mbids
+          ORDER BY cnt DESC
+        """ % (table))
+    rows = query.collect()
+    releases = defaultdict(list)
+    for row in rows:
+        releases[row.user_name].append({
+            'release_name': row.release_name,
+            'release_msid': row.release_msid,
+            'release_mbid': row.release_mbid,
+            'artist_name': row.artist_name,
+            'artist_msid': row.artist_msid,
+            'artist_mbids': row.artist_mbids,
+            'listen_count': row.cnt,
+        })
+    print("Query to calculate release stats processed in %.2f s" % (time.time() - t0))
+    return releases
+
+def main(app_name):
+    """
+    Calculate statistics of users for previous month.
+
+    Args:
+        app_name: Application name to uniquely identify the submitted
+                  application amongst other applications
+
+    """
+    t0 = time.time()
+    try:
+        listenbrainz_spark.init_spark_session(app_name)
+    except Exception as err:
+        logging.error("Cannot initialize spark session: %s / %s. Aborting." % (type(err).__name__, str(err)))
+        sys.exit(-1)
+    date = date_for_stats_calculation()
+    try:
+        df = listenbrainz_spark.sql_context.read.parquet('{}/data/listenbrainz/{}/{}.parquet'.format(config.HDFS_CLUSTER_URI, date.year, date.month))
+        print("Loading dataframe...")
+    except Exception as err:
+        logging.error("Cannot read files from HDFS: %s / %s. Aborting." % (type(err).__name__, str(err)))
+        sys.exit(-1)
+    df.printSchema()
+
+    table = 'listens_{}'.format(datetime.strftime(date, '%Y_%m'))
+    print(table)
+    df.registerTempTable(table)
+    print("Running Query...")
+    query_t0 = time.time()
+    print("DataFrame loaded in %.2f s" % (query_t0 - t0))
+
+    data = defaultdict(dict)
+    #data : Nested dict which can be depicted as:
+            #{'user1' : {'artist' : {artists dict returned by func get_artists},
+            #'recordings' : {recordings dict returned by func get_recordings}, 
+            #'releases': {releases dict returned by func get_releasess}, 'yearmonth' : 
+             #'date when the stats were calculated'}, 'user2' : {...}} 
+
+    artist_data = get_artists(table)
+    for user_name, artist_stats in artist_data.items():
+        data[user_name]['artists'] = {
+            'artist_stats': artist_stats,
+            'artist_count': len(artist_stats),
+        }
+
+    recording_data = get_recordings(table)
+    for user_name, recording_stats in recording_data.items():
+        data[user_name]['recordings'] = recording_stats
+
+    release_data = get_releases(table)
```

Comment:
Here and through the rest of the file, the same try block should be used.

Reply:
After catching exception it is preferable to stop the script because there is a specific format in which data should be sent to LB server, any irregularity in that should terminate the script.